teos/logger.py

```diff
@@ -0,0 +1,225 @@
+
+
+# Convenience method to ignore a signal
+def ignore_signal(_, __):
+    pass
+
+
+def _repr(val):
+    """Returns the representation of *val* if it's not a ``str``."""
+
+    return val if isinstance(val, str) else repr(val)
+
+
+def encode_event_dict(event_dict):
+    """
+    Encodes an event dictionary in a nicely formatted string, following the general format:
+
+        timestamp [component] event (attr1=value1, attr2=value2, ...)
+
+    where values that are not present are omitted. See unit tests for a more precise specification.
+    """
+
+    sio = StringIO()
+
+    ts = event_dict.pop("timestamp", None)
+    if ts:
+        sio.write(str(ts) + " ")
+
+    component = event_dict.pop("component", None)
+    if component:
+        sio.write("[" + component + "] ")
+
+    event = _repr(event_dict.pop("event"))
+
+    sio.write(event)
+
+    # Represent all the key=value elements still in event_dict
+    key_value_part = ", ".join(key + "=" + _repr(event_dict[key]) for key in sorted(event_dict.keys()))
+    if len(key_value_part) > 0:
+        sio.write("  (" + key_value_part + ")")
+
+    return sio.getvalue()
+
+
+class LogRecordStreamHandler(socketserver.StreamRequestHandler):
+    """
+    Handler for a streaming logging request. Sends to the logger any received log message, after some preprocessing.
+    """
+
+    # Taken almost verbatim from Python's logging cookbook.
+    def handle(self):
+        """
+        Handle multiple requests - each expected to be a 4-byte length, followed by the LogRecord in pickle format.
+        Logs the record according to whatever policy is configured locally.
+        """
+
+        while True:
+            chunk = self.connection.recv(4)
+            if len(chunk) < 4:
+                break
+            slen = struct.unpack(">L", chunk)[0]
+            chunk = self.connection.recv(slen)
+            while len(chunk) < slen:
+                chunk = chunk + self.connection.recv(slen - len(chunk))
+            obj = pickle.loads(chunk)
+            record = logging.makeLogRecord(obj)
+            self.handle_log_record(record)
+
+    @staticmethod
+    def handle_log_record(record):
+        """
+        Processes log records received via the socket. The record's ``msg`` field is expected to be an encoded ``dict``
+        produced by structlog. The ``dict`` is encoded to a string using ``encode_event_dict`` and sent to the logger
+        with the name specified in the record.
+
+        Args:
+            record (:obj:`logging.LogRecord`): a log record.
+        """
+
+        event_dict = json.loads(record.msg)
+        message = encode_event_dict(event_dict)
+
+        logger = logging.getLogger(record.name)
+        logger.log(record.levelno, message, exc_info=record.exc_info, stack_info=record.stack_info)
+
+
+class LogRecordSocketReceiver(socketserver.ThreadingTCPServer):
+    """
+    Simple TCP socket-based logging receiver.
+    """
+
+    allow_reuse_address = True
```

Comment:
Is this ever used?

Reply:
I think it needs to be in the class, see for example  https://stackoverflow.com/questions/15260558/python-tcpserver-address-already-in-use-but-i-close-the-server-and-i-use-allow

Reply:
Interesting.